sdk/tests/performance/webgl_webcodecs_video_frame/webgl_webcodecs_video_frame.js

```diff
@@ -0,0 +1,314 @@
+    if (!gl) {
+        console.log("<h1>Unable to initialize WebGL. Your browser or machine may not support it.</h1>");
+        return null;
+    }
+    let ext = gl.getExtension('WebGL_webcodecs_video_frame');
+    if (!ext) {
+        console.log("<h1>Unable to initialize WebGL_webcodecs_video_frame. Your browser or machine may not support it.</h1>");
+        return null;
+    }
+
+    // This helper function generates the required GLSL extension for the video frame.
+    function emitRequiredExtension(videoFrameHandle) {
+        if (videoFrameHandle.requiredExtension !== null) {
+            let str = "";
+            str += "#extension ";
+            str += videoFrameHandle.requiredExtension;
+            str += " : require\n";
+            return str;
+        }
+        return "";
+    }
+
+    // This helper function generates GLSL uniform declarations for the video frame.
+    // Mainly the logic varies according to the pixel format of video frame.
+    //
+    // ABGR 
+    // uniform sampler ${name}_0;
+    //
+    // NV12
+    // uniform samplerExternalOES ${name}_0;
+    // uniform samplerExternalOES ${name}_1;
+    //
+    //  I420
+    // uniform samplerExternalOES ${name}_0;
+    // uniform samplerExternalOES ${name}_1;
+    // uniform samplerExternalOES ${name}_2;
+    function emitSamplerDeclaration(videoFrameHandle, name) {
+        let infoArray = videoFrameHandle.textureInfoArray;
+        let str = "";
+        for (let i = 0; i < infoArray.length; ++i) {
+            str += "uniform " + infoArray[i].samplerType + " _" + name + "_"
+                + i.toString() + "_;\n";
+        }
+        return str;
+    }
+
+    // This helper function generates GLSL samplers to access video frame.
+    //
+    // ABGR
+    // ${out}.rgb = texture2D(${name}_0, ${texCoord}).rgb;
+    //
+    // NV12
+    // ${out}.r = texture2D(${name}_0, ${texCoord}).r;
+    // ${out}.gb = texture2D(${name}_1, ${texCoord}).rg;
+    // 
+    // I420
+    // ${out}.r = texture2D(${name}_0, ${texCoord}).r;
+    // ${out}.g = texture2D(${name}_1, ${texCoord}).r;
+    // ${out}.b = texture2D(${name}_2, ${texCoord}).r;
+    function emitSamplerBody(videoFrameHandle, name, texCoord, out) {
+        let components = [];
+        components["ABGR"] = ["rgb"];
+        components["XRGB"] = ["rgba"];
+        components["NV12"] = ["r", "gb"];
+        components["I420"] = ["r", "g", "b"];
+        let format = videoFrameHandle.pixelFormat;
+        let infoArray = videoFrameHandle.textureInfoArray;
+        let str = "  vec4 " + out + ";\n";
+        for (let i = 0; i < infoArray.length; ++i) {
+            str += "  " + out + "." + components[format][i] + " = " + infoArray[i].samplerFunc
+                + "(" + "_" + name + "_" + i.toString() + "_, "
+                + texCoord + ")." + infoArray[i].components + ";\n";
+        }
+        return str;
+    }
+
+    function emitUnPremultiplyAlpha(videoFrameHandle, pixel) {
```

Comment:
Is un-premultiplication functionality strongly needed? It's lossy, and I'm not sure how good it will look even if the code path is present.

Reply:
MacOS video frames have this pre-alpha stuff. I just follow what CC does currently.